### PR TITLE
Try: Polish link/unlink buttons

### PR DESCRIPTION
**Pull request by jasmussen**
## What?

The link/unlink buttons used across the interface are blue, rectangular, and uses a scaled and blurry/illegible icon:

before

This PR changes those to be plain 24x24 icon buttons toggles:

Screenshot 2022-09-02 at 10 53 34

after

## Why?

Instead of having a separate style just for link/unlink buttons, this focuses in on an existing icon button style already in use, it reduces prominence, and it makes the icon more legible.

## How?

The PR changes the style of the buttons in a few places, and it touches some legacy inherited CSS. It seems like there's an opportunity for a singular link/unlink component, but this at least takes some steps towards it.

## Testing Instructions

Please test border, radius, padding controls and observe the link/unlink button being correctly scaled (24x24) and positioned.

Because this touches on some older styles for "small" buttons, it would be good to look up any instances of _buttons with text that use the isSmall prop_, and see that those still look correct.

### CODE REVIEW

File: packages/block-editor/src/components/border-radius-control/linked-button.js
Comment (jasmussen) [inline packages/block-editor/src/components/border-radius-control/linked-button.js:9]: Sentence case across all of these.

File: packages/block-editor/src/components/border-radius-control/linked-button.js
Comment (jasmussen) [inline packages/block-editor/src/components/border-radius-control/linked-button.js:14]: The extra span, like it does for the other instances of this button, makes the button "not have text", which invokes the `:not(.has-text)` CSS.

File: packages/components/src/border-box-control/border-box-control-linked-button/component.tsx
Comment (jasmussen) [inline packages/components/src/border-box-control/border-box-control-linked-button/component.tsx:33]: In general icons should never be smaller than 24px, with almost no exceptions, because it causes blurry antialiasing of the pixels that were designed for the 24px grid. The only exception I can think of, is the external link icon, which can occasionally be shown in 12px size. But at that size (exact half), the pixels interpolate correctly, so it remains crisp.

File: packages/components/src/border-box-control/styles.ts
Comment (jasmussen) [inline packages/components/src/border-box-control/styles.ts]: Not sure why 22px makes this work. It's there also in the legacy CSS.

File: packages/components/src/button/style.scss
Comment (jasmussen) [inline packages/components/src/button/style.scss:299]: Style changes in this file are probably the most impactful, so it would be good to test _small icon buttons with and without text_.

File: packages/components/src/border-box-control/styles.ts
Comment (aaronrobertshaw) [reply packages/components/src/border-box-control/styles.ts]:
```suggestion
		line-height: 0;
```

> Not sure why 22px makes this work

I'm not sure this does work as `line-height: 22px` increases the button's wrapper height throwing off the overall control height and vertical alignment.

https://user-images.githubusercontent.com/60436221/188401934-ce9d9df3-4ff9-4523-9065-338835c31d1f.mp4

File: packages/components/src/button/style.scss
Comment (aaronrobertshaw) [inline packages/components/src/button/style.scss]: Did you mean to remove the horizontal padding here? I noticed this affects the reset button for the `BoxControl`.

| Before | After |
|---|---|
| Screen Shot 2022-09-05 at 6 06 57 pm | Screen Shot 2022-09-05 at 6 17 49 pm |

Comment (aaronrobertshaw) [reply packages/components/src/button/style.scss]: This also breaks other `isSmall` buttons in the editor such as those for the image block's sizes etc.

Screen Shot 2022-09-05 at 6 19 58 pm

**Review by aaronrobertshaw — CHANGES_REQUESTED** [review]
Thanks for iterating on these buttons @jasmussen 👍 

Unfortunately, I ran into a few issues while testing this PR. I've added inline comments with more details but the TL;DR is:

- `BorderBoxControl` height has been increased due to new styles also throwing off vertical alignment.
- Buttons with `isSmall` and text are missing their horizontal padding e.g. `BoxControl` reset button, Image block sizes etc.
- We also need a changelog for any changes to controls in our components package.

File: packages/components/src/button/style.scss
Comment (jasmussen) [reply packages/components/src/button/style.scss]: Thank you for excellent feedback and thank you for looking. It does seem like this PR is unlikely to land as is.

With this PR in mind as an example, is there a different and safer approach we can take, to ensure the link/unlink button can move forward, i.e. componentization? Or do you still see the iterative approach as taken in this PR as potentially feasible?

Comment (aaronrobertshaw) [reply packages/components/src/button/style.scss]: > With this PR in mind as an example, is there a different and safer approach we can take, to ensure the link/unlink button can move forward, i.e. componentization?

I've actually made the argument internally as well that these linked buttons would be good candidates to make into a reusable component when bandwidth allowed. I don't think doing so moves the redesign of these buttons forward any faster, in fact, likely the opposite. 

> Or do you still see the iterative approach as taken in this PR as potentially feasible?

The iterative approach proposed in this PR makes sense to me as I don't believe it is too far off. 

The style accidentally increasing the `BorderBoxControl`'s height can be fixed easily and with a slight change in approach to tweaking the linked button's padding we should be very close to landing this.

Comment (jasmussen) [reply packages/components/src/button/style.scss]: Cool thank you, I'll see if I can polish this along as I get a breather! 🙏

**Comment by jameskoster** [comment]
Should the icon be the other way around?

IE, when the sides are linked, the button label should be the _unlink_ icon since that describes what will happen on click?

**Comment by jasmussen** [comment]
Showing the full link when sides are linked seems to be the standard in Photoshop, Illustrator, and Figma:

figma

**Comment by jasmussen** [comment]
Alright, thanks to excellent sleuthing by @aaronrobertshaw, I think this one is ready for another review:

status

**Review by jameskoster — APPROVED** [review]
https://user-images.githubusercontent.com/846565/190122062-3e3f6109-fea9-4ade-884b-f4bf2c927726.mp4

Much better, thank you!

**Comment by jasmussen** [comment]
Cool, I'll wait for Aaron to sanity check and then land this! 🙏

File: packages/components/src/button/style.scss
Comment (ciampo) [reply packages/components/src/button/style.scss:299]: cc @mirka and @t-hamano (related to #44042)

File: packages/block-editor/src/components/border-radius-control/linked-button.js
Comment (ciampo) [reply packages/block-editor/src/components/border-radius-control/linked-button.js:14]: I'm not sure I follow here — how can a parent `span` influence the `has-text` classname on the `Button` component?

File: packages/components/src/button/style.scss
Comment (t-hamano) [reply packages/components/src/button/style.scss:299]: This CSS seems to affect buttons that have `isSmall` and `icon` prop and no text.

For one, it changes the width from **36px** to **24px**.
This is because the `trunk` applies `min-width: 36px;` even if the button is small:

https://github.com/WordPress/gutenberg/blob/9a345b7e5c03ea57042115dcccb0bca0c0764d30/packages/components/src/button/style.scss#L305-L307

Second, the SVG icon size is different. In the `trunk`, the SVG icons are squashed into 8px padding on the left and right, and displayed at **20px**. (36px - 8px * 2 = 20px)
In this PR, the icon size is **24px**.

However, I think this is not the intended behavior in trunk, and the change made by this PR is expected behavior.

### Trunk

trunk

### This PR

pr

File: packages/block-editor/src/components/border-radius-control/linked-button.js
Comment (aaronrobertshaw) [reply packages/block-editor/src/components/border-radius-control/linked-button.js:14]: I'm also unclear as to why this `span` wrapper is required. It looks like it's been copied from the approach in the `BoxControl` or because the new spacing presets UI needed one? Removing the wrapper didn't appear to make any visual change to me.

| With `span` | Without `span` |
|---|---|
| Screen Shot 2022-09-15 at 4 43 04 pm | Screen Shot 2022-09-15 at 4 43 42 pm |

If it isn't required, I'd vote to clean that up as well. Seeing if the other components (spacing presets and box control) still need the spans wrapping their buttons could be a follow-up though.

**Review by aaronrobertshaw — COMMENTED** [review]
Thanks for iterating on this PR @jasmussen 👍 

It looking and testing better for me now. 

✅ Issue with increased `BorderBoxControl` height has been resolved
✅ Horizontal padding has been restored to `isSmall` buttons with text e.g. width buttons
✅ Couldn't find any other instances of isSmall buttons that looked off
❓ There appears to be an unnecessary `span` wrapping the border radius button as Marco [highlighted](https://github.com/WordPress/gutenberg/pull/43802/files#r970902531)

If we can remove the span noted above I think this is almost ready to merge. It might pay to double check with @t-hamano regarding the icon sizing behaviour he [raised](https://github.com/WordPress/gutenberg/pull/43802/files#r970971892) and confirm we're happy to proceed.



https://user-images.githubusercontent.com/60436221/190338366-940c54a3-06ac-4473-953a-51fe3b1eedd2.mp4

File: packages/block-editor/src/components/border-radius-control/linked-button.js
Comment (jasmussen) [reply packages/block-editor/src/components/border-radius-control/linked-button.js:14]: Inded that's curious. It was definitely needed at one point, but it doesn't appear to be anymore. Removed!

File: packages/components/src/button/style.scss
Comment (jasmussen) [reply packages/components/src/button/style.scss:299]: Thank you for looking, @t-hamano, much appreciated. Inded, the intended behavior is for the icon to always be 24x24. The screenshot of this branch above looks like the correct behavior to me.

**Comment by jasmussen** [comment]
Thank you, I removed the span and it appears to behave as intended still:

without span

I would appreciate a final look from @t-hamano and @ciampo if either has time. Thanks everyone!

File: packages/block-editor/src/components/border-radius-control/linked-button.js
Comment (jasmussen) [reply packages/block-editor/src/components/border-radius-control/linked-button.js:14]: Actually my mistake, I removed it only from one of them. When I removed it from the border version, this happened:

without span

Let me look at a different solution than adding a span.

Comment (jasmussen) [reply packages/block-editor/src/components/border-radius-control/linked-button.js:14]: Oh, it's because for some reason, just the border-radius-link button gets a `has-text` CSS class applied, even though it shouldn't. 🤔

Comment (jasmussen) [reply packages/block-editor/src/components/border-radius-control/linked-button.js:14]: Yeah I actually don't know how to fix this. When the span is present, the `has-text` class is not. When the span is removed, `has-text` appears. And that means the following icon-button-CSS doesn't apply anymore:

```
	.components-button.is-small.has-icon:not(.has-text) {
		min-width: $icon-size;
		padding: 0;
	}
```

Comment (jasmussen) [reply packages/block-editor/src/components/border-radius-control/linked-button.js:14]: @ciampo if you have any insights, I'd appreciate it. This appears to be the button piece that adds the problematic class:

```
		'has-text': !! icon && !! children,
```

Feel free to push to the branch in case you have a fix.

Comment (t-hamano) [reply packages/block-editor/src/components/border-radius-control/linked-button.js:14]: I was just looking into this.
Curiously, if the `Button` component is directly placed under the `ToolTip`, it has `null` children.

check

This would cause it to be considered as having text.
Inserting a span tag would solve this problem, but I'll look into it some more.

Comment (jasmussen) [reply packages/block-editor/src/components/border-radius-control/linked-button.js:14]: Thank you 🙏

Comment (t-hamano) [reply packages/block-editor/src/components/border-radius-control/linked-button.js:14]: @ciampo 

I believe this is probably `Tooltip` merging refs.
This may be a problem that should be resolved on the `Tooltip` component, but I have tried an approach to check if the child is empty on the `Button` component.

This change resolves the issue, but is this appropriate?

```diff
--- a/packages/components/src/button/index.js
+++ b/packages/components/src/button/index.js
@@ -92,6 +92,8 @@ export function Button( props, ref ) {
                'components-button__description'
        );
 
+       const hasChildren = children?.[ 0 ] && children[ 0 ] !== null;
+
        const classes = classnames( 'components-button', className, {
                'is-secondary': variant === 'secondary',
                'is-primary': variant === 'primary',
@@ -101,7 +103,7 @@ export function Button( props, ref ) {
                'is-busy': isBusy,
                'is-link': variant === 'link',
                'is-destructive': isDestructive,
-               'has-text': !! icon && !! children,
+               'has-text': !! icon && hasChildren,
                'has-icon': !! icon,
        } );
```

Comment (ciampo) [reply packages/block-editor/src/components/border-radius-control/linked-button.js:14]: > @ciampo if you have any insights, I'd appreciate it. This appears to be the button piece that adds the problematic class:

I don't think I've ever really worked on `Button` since I started working on components, so I'd need to look into it specifically to understand what's going on.

Although the fix proposed above seems reasonable! @t-hamano, would you be able to open a PR with your suggested code changes and ping me there, so that we can assess them and hopefully merge the fix?

Comment (ciampo) [reply packages/block-editor/src/components/border-radius-control/linked-button.js:14]: In that new PR, we could even include [these changes](https://github.com/WordPress/gutenberg/pull/43802#discussion_r970971892) — so that we keep that PR focused on the `Button` component

Comment (jasmussen) [reply packages/block-editor/src/components/border-radius-control/linked-button.js:14]: Should we just include it in this PR? Otherwise can we land this one _with_ the span intact (PR as-is now), and then remove the span in that new separate PR?

Comment (jasmussen) [reply packages/block-editor/src/components/border-radius-control/linked-button.js:14]: I'm asking because it's pretty crucial we get this PR merged before the freeze, so anything we can do to land this one sooner rather than later would be great. The "has-text" issue seems like a bug that can be fixed after the freeze as well.

Comment (t-hamano) [reply packages/block-editor/src/components/border-radius-control/linked-button.js:14]: I will begin work on a PR for the button.
If @jasmussen's PR should be merged first, I think there is no problem with reinstating the `span` tag.
I think it's not too late to decide whether to remove `span` tag or not after the two PRs have been merged in the end.

Comment (t-hamano) [reply packages/block-editor/src/components/border-radius-control/linked-button.js:14]: I have submitted #44198, but I need to do further validation because some unit tests are failing 😅
I vote to proceed with this well-reviewed PR merge.

**Review by mirka — APPROVED** [review]
Jumping in because Marco has wrapped up for the week — I'm also fine with merging this PR with the temporary `span` in place 👍 Maybe add a `// TODO: Remove span after merging https://github.com/WordPress/gutenberg/pull/44198` or something like that so we don't forget.

We should also add a changelog for [this size change](https://github.com/WordPress/gutenberg/pull/43802/files?w=1#r971839175) because I imagine a lot of consumers have added their own style overrides to fix this.

**Comment by jasmussen** [comment]
Thank you!

I've kept the spans in place, but added a comment. @aaronrobertshaw if you want a last look, I'd appreciate it, otherwise I'm going to land this one when the tests pass.

**Review by aaronrobertshaw — APPROVED** [review]
@jasmussen Thanks for the ping and seeing this one through 👍 

I've given this a quick test and it's working well for me. 

We still need the a [changelog](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/CHANGELOG.md) as noted in a [few](https://github.com/WordPress/gutenberg/pull/43802#pullrequestreview-1096038999) earlier [comments](https://github.com/WordPress/gutenberg/pull/43802#pullrequestreview-1109676301) before we merge this one though. Just a  quick note detailing the changes to the `BorderBoxControl`, `BoxControl`, and `Button` components is all that's needed.

Once that changelog is in place and tests are all green, merge away! 🚢

**Comment by jasmussen** [comment]
Ack my apologies, I forgot to update the changelog (long week). Done now, let me know if it needs changing. Thanks everyone, and I'm happy to land this. 🙏

**Comment by ockham** [comment]
I just cherry-picked this PR to the wp/6.1 branch to get it included in the next release: 810ba68ca6

**Pull request merged**
